// office-ui-fabric-react/lib-amd/demo/pages/Facepile/examples/Facepile.Basic.Example.js
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
define(["require", "exports", 'react', '../../../../index'], function (require, exports, React, index_1) {
    "use strict";
    var facepileProps = {
        personas: [
            {
                personaName: 'Annie Lindqvist',
                imageUrl: './images/persona-female.png'
            },
            {
                personaName: 'Greta Lundberg',
                imageInitials: 'GL',
                initialsColor: index_1.PersonaInitialsColor.green
            },
            {
                personaName: 'Roko Kolar',
                imageInitials: 'RK',
                initialsColor: index_1.PersonaInitialsColor.purple,
                data: 'Emp1234',
                onClick: function (ev, persona) {
                    return alert('You clicked on ' + persona.personaName + '. Extra data: ' + persona.data);
                }
            }
        ]
    };
    var FacepileBasicExample = (function (_super) {
        __extends(FacepileBasicExample, _super);
        function FacepileBasicExample() {
            _super.apply(this, arguments);
        }
        FacepileBasicExample.prototype.render = function () {
            return (React.createElement(index_1.Facepile, __assign({}, facepileProps)));
        };
        return FacepileBasicExample;
    }(React.Component));
    exports.FacepileBasicExample = FacepileBasicExample;
});
